support vim replace command with range

cthulhua opened the pull request:
Release Notes:

- Added support for line ranges in vim replace commands #9428


- not supporting anything other than bare line numbers right now
- ~need to figure out how to show range in question in search bar~ @ConradIrwin implemented showing a highlight of the selected range for a short direction instead
- ~tests lol~

cthulhua committed:
allow vim replace with a range

WIP

cthulhua committed:
Merge branch 'zed-industries:main' into cthulhua/vim-replace-range

cthulhua commented:
@cla-bot check

ConradIrwin commented:
@cthulhua if you'd like to work on this together for a bit, please book time here: https://calendly.com/conradirwin/pairing.

With the caveat that I haven't tried to implement any of this, here are some ideas from looking at the change:

For UX there are a few ways we could go here:
* Build it into search (possibly conditional on vim mode)
* Hide it, and keep it as a feature of command (as vim itself does)

I'm leaning towards the second option. This avoids adding fiddly UI for a relatively rare use-case, and hopefully simplifies building the feature. It would also make it consistent with using ranges in things like `:2,5j`, where there's no real UI for it – it just happens when you hit enter. It will make making `<up>` work in the command palette more urgent, but that can be a separate PR.

If we go down this route, I think we could descope any UI changes, and make a change just to `search_bar.replace_all`. Maybe it could take an optional callback that determines whether a given match should be replaced? Then we could implement `/g` using the same mechanism?

We should still deploy the existing UI with the search and replace, but as the range would not be stateful, there's no need to display it.

cthulhua committed:
Merge branch 'zed-industries:main' into cthulhua/vim-replace-range

cthulhua committed:
refactor wrt ranges, remove search bar stuff

cthulhua committed:
move range regex to `OnceLock` in search.rs

instead of a `lazy_static`, let's use `OnceLock`. Also move to the
search module, as that's more tightly coupled to the regex. We have to
add a `range_regex()` fn that does the init, because I really don't want
to init this from two different places. `LazyLock` will be nice whenever
it's stabilized.

cthulhua committed:
remove some line noise

inadvertently introduced in previous impl

cthulhua commented:
@ConradIrwin got rid of that lazy_static, I think that was the only thing we'd wanted to do
